### DISCUSSION TITLE
Question about sensor data used by firmware.

### DISCUSSION THREAD

**Discussion by dixxLee** [bitcraze/discussions#491] [Q&A] [answered]
Hello.
I'm a master's student that doing research using Crazyflie 2.1 at Sejong Univ. South korea.
While analyzing Crazyflie firmware, I have a question because I didn't understand something.

In 'stabilizer.c', There are 2 structures in this c file : sensorData_t and state_s.
The part that I want to focus on here is the accel data.

The code describes the state data as "but acc.z without considering gravity".
I don't understand this.

I wonder how the data of 'sensorData_t' gets the data of the sensor, and how the data of 'state_s' gets the data of the sensor. 
I also wonder if it goes through filtering.
In addition, in the case of the accel data of state_s, it seems that the data of x, y, and z in the free fall state converges to almost 0, and I wonder if this data shows the measured data of the sensor as a specific force.

![image](https://user-images.githubusercontent.com/46697603/209774376-1ce46640-0958-45e9-978c-61fde3186aee.png)

Please explain the difference between 'sensorData_t' and 'state_s'.

Thanks.
Happy New Year.

**Comment by knmcguire** [comment, marked as answer]
The difference between the sensor struct and state struct is as follows:
- the sensor struct contains the sensor data that comes directly from the sensor drivers as well, either raw or slightly filtered by some setting
- the state struct contains the estimate values of the state, which is not the direct sensor data but results from sensor fusion and a model of the system. 

I would refer to this documentation to learn more about the stabilizer, sensor control state system of the crazyflie, that will clear things up: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/

I guess for the acc.z. the gravity factor (9.81 m/s2) has been removed, so it is only the acceleration of the quadcopter itself.

**Reply by dixxLee** [reply]
I am very grateful for your reply.
Thanks.
